File: combina_niche3.m
function [results1]=combina_niche3(sp_chr,T)
% Needs 'niche_difer_sp.m' and 'niche_difer2.m'
% INPUTS:
%
% sp_chr: a matrix with the species chromatograms (categories by environmental variables by species). Outputs of 'chromato_env16.m'
% T: the threshold of minimal abundance in a category for niche breadth estimation
%
% OUTPUTS:
%
% results1: The mean degree of niche overlapping. The first column displays the number of dimensions 
% considered simultaneously, columns 2 to 10 display the combinations of dimensions. 
% The last column displays index D associated with the combination of environmental dimensions. 
% D=0 when species niches are fully different and D=100 when species niches are identical; 
% the higher the number of dimensions, the lower the value of index D. 
% Only the combinations of environmental variables that minimise values of index D are displayed.
%
% Grégory Beaugrand Avril 2021

[~,p,z]=size(sp_chr);        % CATEGORIES X DIMENSIONS X ESPECES
v=[1:p];
nb=0;
for i=1:p
    clear temp ntemp ptemp
    temp=combnk(v,i);
    [ntemp,~]=size(temp);
    nb=ntemp+nb;
end
clear temp ntemp ptemp

point=zeros(nb,1)+nan;
point2=zeros(nb,p)+nan;
y=zeros(z,z,nb)+nan;
moyenne=zeros(nb,1)+nan;

cp=0;
for i=1:p
    clear temp ntemp ptemp 
    temp=combnk(v,i);
    [ntemp,~]=size(temp);
    for j=1:ntemp
        cp=cp+1;
        y(:,:,cp)=niche_difer_sp(sp_chr(:,temp(j,:),:),T);
        point(cp,1)=i;
        point2(cp,1:i)=temp(j,1:i);
        
        clear test test2 test3 temp2
        test=ones(z,z);test2=triu(test,1);
        test3=squeeze(y(:,:,cp));
        temp2=test3(test2==1);
        moyenne(cp,1)=mean(temp2,'omitnan');
    end
end

results1=nan(p,p+2);

for i=1:p
    clear f temp1
    f=find(point==i);
    temp1=moyenne(f,1);
    [mini1,val1]=min(temp1,[],1,'omitnan');
    results1(i,:)=[i point2(f(val1),:) mini1];
end

File: combina_niche3.R
combina_niche3<-function(sp_chr,T){
  source("niche_difer_sp.R")
  n<-dim(sp_chr)
  v<-(1:n[2])
  nb<-0
  
  for (i in 1:n[2]) {
    temp<-t(combn(v,i))
    ntemp<-dim(temp)
    nb<-ntemp[1]+nb
    rm(temp,ntemp)
  }
  
  
  point<-matrix(nrow = nb,ncol = 1)
  point2<-matrix(nrow = nb,ncol = n[2])
  y<-array(data=NA,dim = c(n[3],n[3],nb))
  moyenne<-matrix(nrow = nb,ncol = 1)
  
  cp<-0
  for (i in 1:n[2]) {
    temp<-t(combn(v,i))
    ntemp<-dim(temp)
    
    for (j in 1:ntemp[1]) {
      cp<-cp+1
      y[,,cp]<-niche_difer_sp(sp_chr[,temp[j,],],T)
      point[cp]<-i
      point2[cp,1:i]<-temp[j,1:i]
      
      test<-y[,,cp]
      test2<-upper.tri(test,diag = FALSE)
      moyenne[cp]<-mean(test[test2==TRUE],na.rm = TRUE)
      rm(test,test2)
    }
    rm(temp,ntemp)
  }
  
  results1<-matrix(nrow = n[2],ncol = n[2]+2)
  
  for (i in 1:n[2]) {
    f<-which(point==i)
    temp1<-moyenne[f]
    mini<-min(temp1,na.rm = TRUE)
    pos<-which(temp1==mini)
    results1[i,]<-t(c(i,point2[f[pos],],mini))
    rm(f,temp1,mini,pos)
  }
  results1<-results1
}
